Add Solana support

0xmichalis opened the pull request:
Add support for labeling Solana addresses, which can be between 32-44 characters
and are Base58-encoded (using charset [1-9A-HJ-NP-Za-km-z]). Supports:
- Standard wallet addresses (32-byte public keys)
- Program Derived Addresses (PDAs)
- Program addresses (smart contracts)
- System Program address (32 ones)
- Associated Token Account addresses

Related issue: https://github.com/rolod0x/rolod0x/issues/419

0xmichalis committed:
refactor: move single line parsing into its own function

0xmichalis committed:
refactor: move EVM-specific logic into its own function

0xmichalis committed:
feat: add Solana support

Add support for labeling Solana addresses, which can be between 32-44 characters
and are Base58-encoded (using charset [1-9A-HJ-NP-Za-km-z]). Supports:

- Standard wallet addresses (32-byte public keys)
- Program Derived Addresses (PDAs)
- Program addresses (smart contracts)
- System Program address (32 ones)
- Associated Token Account addresses

0xmichalis committed:
refactor: share line parts parsing logic

0xmichalis committed:
refactor: update isAbbreviation to support multiple patterns

0xmichalis committed:
test: fix test description

aspiers commented on src/shared/regexps.ts:
```diff
@@ -1,3 +1,13 @@
+// EVM regexps
 export const RE_ADDRESS = /0x[0-9a-f]{40}\b/i;
```
Should probably rename this to `RE_EVM_ADDRESS` for clarity.

0xmichalis replied:
Addressed in https://github.com/rolod0x/rolod0x/pull/490/commits/d5db96d1a9a550dc39d8c31a82100ca86634e708

aspiers commented on src/shared/abbreviators.ts:
```diff
@@ -45,16 +45,44 @@ const ABBREVIATION_LENGTHS = [
+  [4, 4],
+];
+
 export function abbreviatedAddresses(address: string): string[] {
```
Probably should rename to `abbreviatedEVMAddresses`.

0xmichalis replied:
Addressed in https://github.com/rolod0x/rolod0x/pull/490/commits/d5db96d1a9a550dc39d8c31a82100ca86634e708

aspiers commented on src/shared/abbreviators.ts:
```diff
@@ -45,16 +45,44 @@ const ABBREVIATION_LENGTHS = [
   [30, 0],
 ];
 
+const ABBREVIATION_LENGTHS_SOLANA = [
```
Probably should rename `ABBREVIATION_LENGTHS` to `ABBREVIATION_LENGTHS_EVM`.

0xmichalis replied:
Addressed in https://github.com/rolod0x/rolod0x/pull/490/commits/d5db96d1a9a550dc39d8c31a82100ca86634e708

aspiers commented:
Looks great! Yes you are right the 0x prefix handling needs changing.  I wonder if maybe we should generalise that to `%p` which formats as `0x` for EVM replacements and as the empty string for Solana replacements.  That would also allow future support for many other prefixes in use (e.g. `bc1` for SegWit).  Thoughts?

0xmichalis commented:
> That would also allow future support for many other prefixes in use (e.g. `bc1` for SegWit). Thoughts?

Presumably, Bitcoin support would also include legacy (`1..`) and script (`3...`) formats so such a feature may not be reusable for Bitcoin, unless you choose to only support the newer formats only. 

More general though, and mind you I haven't spent a lot of time thinking about any of this:
* I was thinking that instead of a single format pattern, we probably need a pattern per network as you can have an addressbook that contains addresses from multiple networks but you may want to apply different rules depending on the network of the address(es) that are getting formatted at any point in time. This problem becomes immediately obvious in the CLI where there is only a single format option (`-f`) but a file may contain addresses from multiple networks and EVM's current rule of prefixing `0x` breaks Solana addresses (at least that's how I have been testing my changes so far).
* May also be worth to consider whether you want to allow mixing addresses from multiple networks in a single file. This would certainly make it easier for the CLI but in the backend code, you likely still need to be able to format addresses from multiple networks at the same time so formatting per network is still the way to go?

aspiers commented:
[**@0xmichalis**](https://github.com/0xmichalis) commented on [January 22, 2025 11:43 AM](https://github.com/rolod0x/rolod0x/pull/490#issuecomment-2607022364 "Wed Jan 22 2025 11:43:23 GMT+0000 - Replied by Github Reply Comments"):
> > That would also allow future support for many other prefixes in use (e.g. `bc1` for SegWit). Thoughts?
> 
> Presumably, Bitcoin support would also include legacy (`1..`) and script (`3...`) formats so such a feature may not be reusable for Bitcoin, unless you choose to only support the newer formats only.

Why not?  `%p` would be replaced by whichever prefix was in the original; it wouldn't have to be a single fixed value per chain.

> More general though, and mind you I haven't spent a lot of time thinking about any of this:
> * I was thinking that instead of a single format pattern, we probably need a pattern per network as you can have an addressbook that contains addresses from multiple networks but you may want to apply different rules depending on the network of the address(es) that are getting formatted at any point in time.

Can you give any examples of when it might be useful to apply different formatting per network?  My concern is that this can suddenly multiply the complexity of the display options, which are already far too technical for most users.  E.g. if we have support for 5 different chains then suddenly we're talking about 10 format strings instead of 2.  This sounds like poor UX to me.  The only way I can imagine this being kept manageable is if there's a way to derive format strings from all chains from one base setting, with the option to override per chain.  But unless there's a compelling use case, I'd much prefer to dodge all that complexity, and allowing format strings like `%p%4l...%4r` feels like it should work for all chains pretty nicely.

> This problem becomes immediately obvious in the CLI where there is only a single format option (`-f`) but a file may contain addresses from multiple networks and EVM's current rule of prefixing `0x` breaks Solana addresses (at least that's how I have been testing my changes so far).

Right, but `%p` solves that doesn't it?

> * May also be worth to consider whether you want to allow mixing addresses from multiple networks in a single file.

I definitely do, otherwise this is an artificial UX constraint which will just get in the way of users. 

> This would certainly make it easier for the CLI but in the backend code, you likely still need to be able to format addresses from multiple networks at the same time so formatting per network is still the way to go?

Probably depends on what you mean by "formatting per network".  I don't yet see the need for format _strings_ per network, but to support `%p` the formatter will have to know which network the address is from.  That means the network will need to be passed through from the parser to the mapper, and from there to the formatter.  This should be very easy.

0xmichalis commented:
@aspiers rebased, should be ready for another pass

0xmichalis committed:
docs: add GitHub guidelines in AGENTS.md

0xmichalis committed:
fix: make EVM regexps more robust

This change removes /i from the EVM regexps
and makes them case-insensitive purely based on
the regexp pattern. This makes it easier to
combine these regexps with case-sensitive
regexps that other languages may require.

Also a case is fixed where EVM addresses
preceded by whitespace were not discovered
correctly.

0xmichalis commented on src/shared/parser.ts:
```diff
@@ -30,43 +31,84 @@ export class Parser {
+
+    const isSolanaAddress = isAddress(address);
+    if (isSolanaAddress) this.addEntry(address, label, comment);
+    return isSolanaAddress;
```
applied

0xmichalis commented on src/shared/parser.ts:
```diff
@@ -30,43 +31,84 @@ export class Parser {
+  }
+
+  private parseEVMLine(line: string, lineIndex: number): boolean {
+    const labelLineRe = /^s*(0x[\da-f]{40})\s+(.+?)(?:\s+\/\/\s*(.*?)\s*)?$/i;
```
fixed

0xmichalis commented on src/shared/regexps.ts:
```diff
@@ -1,3 +1,20 @@
+  'i',
+);
+export const RE_ADDRESSES_OR_BYTES32 = new RegExp(
+  `(${RE_EVM_ADDRESS.source}|${RE_EVM_BYTES32.source}|${RE_SOLANA_ADDRESS.source})`,
+  'i',
+);
```
removed

0xmichalis commented on src/shared/regexps.ts:
```diff
@@ -1,3 +1,20 @@
+
+// Combined regexps
+export const RE_ADDRESS_FORMATS = new RegExp(
+  `(${RE_EVM_ADDRESS.source}|${RE_SOLANA_ADDRESS.source})`,
+  'i',
+);
```
removed

0xmichalis commented on src/shared/parser.ts:
```diff
@@ -30,43 +31,84 @@ export class Parser {
 
-    return this.parsedEntries;
+  private parseSolanaLine(line: string, lineIndex: number): boolean {
+    const labelLineRe = /^([1-9A-HJ-NP-Za-km-z]{32,44})\s+(.+?)(?:\s+\/\/\s*(.*?)\s*)?$/;
```
fixed

0xmichalis commented on src/shared/formatter.ts:
```diff
@@ -16,11 +16,23 @@ export class Formatter {
+  // Slice the address from the start to the given number of digits.
+  // If the address starts with 0x, add it back to the sliced address.
+  sliceAddressStart(address: Address, digits: number): string {
+    let addressString = address.replace(/^0x/, '').slice(0, Number(digits));
```
updated

0xmichalis commented on src/shared/abbreviators.ts:
```diff
@@ -69,32 +89,62 @@ export function krakenAbbreviation(address: string): string {
+
+const isSolanaAbbreviation = (abbreviation: string, fullAddress: string): boolean => {
+  const solanaMatch = abbreviation.match(
+    /^([1-9A-HJ-NP-Za-km-z]+)(?:\.\.\.([1-9A-HJ-NP-Za-km-z]+))?$/i,
```
removed

0xmichalis commented on src/pages/content/contextMenu.ts:
```diff
@@ -15,7 +15,7 @@ async function addLabelForClickedElement(): Promise<void> {
   let url = IFRAME_URL;
 
-  const match = clickedEl.outerHTML?.match(RE_ADDRESS);
+  const match = clickedEl.outerHTML?.match(RE_EVM_ADDRESS);
```
fixed

0xmichalis commented on src/shared/mapper.ts:
```diff
@@ -46,8 +47,9 @@ export class Mapper {
+    for (const func of ABBREVIATION_FUNCTIONS[addressType]) {
       for (const abbrev of func(address)) {
         this.labelMap.set(abbrev, guess);
         this.labelMap.set(abbrev.toLowerCase(), guess);
```
fixed

0xmichalis commented on src/shared/formatter.test.ts:
```diff
@@ -1,42 +1,53 @@
-  it('formats a label combining internal digits with other formats', () => {
-    // Test the format used by Kraken
-    const formatter = new Formatter('0x%2l %2i4 ... %-8i4 %4r');
-    expect(formatter.format(label, addr)).toEqual('0xe3 D823 ... D62e F109');
+  const testCases = [
+    { name: 'evm address with 0x prefix', addr: '0xe3D82337F79306712477b642EF59B75dD62eF109' },
+    { name: 'evm address without 0x prefix', addr: 'e3D82337F79306712477b642EF59B75dD62eF109' },
+  ];
```
tests added

0xmichalis commented on src/cli.ts:
```diff
@@ -79,7 +79,7 @@ function getParser(addressesFile: string): Parser {
   const mapper = getMapper(addressesFile, options);
   const rl = readline.createInterface({ input: process.stdin });
-  const regexp = new RegExp(RE_ADDRESS_OR_BYTES32.source, 'gi');
+  const regexp = new RegExp(RE_ADDRESSES_OR_BYTES32.source, 'gi');
```
removed

0xmichalis committed:
feat: add a few more solana default sites

0xmichalis committed:
fix: preserve Solana label formatting behavior

Keep linked-address replacement case-sensitive for Solana so trusted
link targets do not suppress valid text-node matches.

Migrate legacy default display formats so existing users stop rendering
non-EVM addresses with an incorrect 0x prefix.

Made-with: Cursor
